### ISSUE TITLE
Constant error with Chat History

### ISSUE THREAD

**Issue by jhaywoodcobranet** [microsoft/sample-app-aoai-chatGPT#358]
If I enable chat history I get an error after each input. CosmoDB looks ok and I can see the history being added to the container.

![Capture](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/78652044/6611bd53-f561-4e6a-af60-4edb4fbb45f4)

**Comment by jhaywoodcobranet** [comment]
![image](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/78652044/55b84a53-0e16-4b0b-9fbb-6572145020ab)

**Comment by meyerf99** [comment]
I noticed the same thing with the latest Main branch. 

Information is stored in the MongoDB but the error message supposedly prevents the call to the history and no more results are returned.

**Comment by meyerf99** [comment]
@sarah-widder & @jhaywoodcobranet 
The problem could be solved temporarily by deploying commit https://github.com/microsoft/sample-app-aoai-chatGPT/commit/6e644c151d097b80490b340aecee4050f5bb0ac7.

It looks like there is a problem in one of the following two commits.
- https://github.com/microsoft/sample-app-aoai-chatGPT/commit/edcc5b890a4eb516d0c94780fb02b8e816a0e395
- https://github.com/microsoft/sample-app-aoai-chatGPT/commit/4ef422f04c628568cfffe86a8688d3d7933468bb

**Comment by nrobert** [comment]
Same findings here.

Looks like error is in the "update" call:
![image](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/6305666/e5accd53-a553-4c2d-8322-0f1664adbdcd)

The payload has incorrect items (see items 4, 5, 7 and 8 here) which breaks later processing here: https://github.com/microsoft/sample-app-aoai-chatGPT/blob/main/app.py#L523C1-L523C1

**Comment by mvvgit** [comment]
Root cause seems to be that toolMessage in makeApiRequestWithCosmosDB of file Chat.tsx is empty which causes the following call to insert an empty object: 

resultConversation.messages.push(toolMessage, assistantMessage); 

Note that there are 2 instances of the aforementioned code in that file. Both need to be corrected.

Something like the below will fix that:
                    isEmpty(toolMessage) ?
                    resultConversation.messages.push(assistantMessage) :
                    resultConversation.messages.push(toolMessage, assistantMessage);

**Comment by jimmygitctrl** [comment]
> Root cause seems to be that toolMessage in makeApiRequestWithCosmosDB of file Chat.tsx is empty which causes the following call to insert an empty object:
> 
> resultConversation.messages.push(toolMessage, assistantMessage);
> 
> Note that there are 2 instances of the aforementioned code in that file. Both need to be corrected.
> 
> Something like the below will fix that: isEmpty(toolMessage) ? resultConversation.messages.push(assistantMessage) : resultConversation.messages.push(toolMessage, assistantMessage);

I have the same issue ("An error occurred. Answers can't be saved at this time. If the problem persists, please contact the site administrator."), viewing the Log Stream while trying to chat throws this:
  ERROR:root:Exception in /history/update
  Traceback (most recent call last):
    File "/tmp/8dbe1253ac457df/app.py", line 524, in update_conversation
      if len(messages) > 1 and messages[-2]['role'] == "tool":
                               ~~~~~~~~~~~~^^^^^^^^
  KeyError: 'role'

Since I'm deploying via the Azure OpenAI Studio, I can't just modify the code in wwwroot to use the suggested fix (when you restart the Web App it gets overwritten with whatever is inside the /home/site/wwwroot output.tar.gz file which I believe gets extracted to run from /tmp due to oryx being used, and I believe that .tar.gz file is generated based on the contents in this GitHub repo).

Any chance we can get this fixed in the main GitHub so deployments via Azure OpenAI Studio will work with chat history?  Core feature we need to work (compliance says we need to record the questions asked, and ideally, the answers AI returns).

**Comment by iseabock** [comment]
Hello all, 

A quick fix for this issue has been merged with a more robust solution to follow. Apologies for the inconvenience!

**Comment by jimmygitctrl** [comment]
> Hello all,
> 
> A quick fix for this issue has been merged with a more robust solution to follow. Apologies for the inconvenience!

Has it been merged into the main repository?  Otherwise how can someone deploying via the Azure OpenAI Studio pull in this change?  I just re-deployed the app now and the bug is still there.

**Comment by sarah-widder** [comment]
Fixed in #385. Closing the issue
